### PR TITLE
[REVIEW] gdf_group_by_without_aggregations returns gdf_column close #2239

### LINKED ISSUE #2239: [FEA] `gdf_group_by_without_aggregations` should return a column instead of device_vector for indices
Talked offline with @jrhemstad and @harrism and we agreed this allows for the most composability in using the library. This is a blocker for #1747.

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
gdf_group_by_without_aggregations was returning device_vector and it had to return column, so the current set of changes will return gdf_column, once cudf column is available, it will be changed to it close #2239 .

### CODE REVIEW

File: cpp/include/cudf/groupby.hpp [hunk 1]
```diff
@@ -191,10 +191,10 @@ gdf_unique_indices(cudf::table const& input_table, gdf_context const& context);
  * @returns A tuple containing:
  *          - A cudf::table containing a set of columns sorted by the key
  * columns.
- *          - A device vector containing the first index of every unique row
+ *          - A GDF column containing the first index of every unique row
```
Comment (jrhemstad) [inline cpp/include/cudf/groupby.hpp]: Specify the type and nullability.

File: cpp/src/groupby/new_groupby.cu [hunk 2]
```diff
@@ -269,17 +271,28 @@ gdf_unique_indices(cudf::table const& input_table, gdf_context const& context)
...
-std::pair<cudf::table, rmm::device_vector<gdf_index_type>> 
+std::pair<cudf::table, gdf_column>
 gdf_group_by_without_aggregations(cudf::table const& input_table,
                                   gdf_size_type num_key_cols,
                                   gdf_index_type const * key_col_indices,
                                   gdf_context* context)
 {
   CUDF_EXPECTS(nullptr != key_col_indices, "key_col_indices is null");
   CUDF_EXPECTS(0 < num_key_cols, "number of key colums should be greater than zero");
+  gdf_column unique_indices;
```
Comment (jrhemstad) [inline cpp/src/groupby/new_groupby.cu]: You don't need to manually zero out all of the members if you use this form of construction. 
```suggestion
  gdf_column unique_indices{};
```

File: cpp/src/groupby/new_groupby.cu [hunk 3]
```diff
@@ -350,6 +363,12 @@ gdf_group_by_without_aggregations(cudf::table const& input_table,
   std::transform(key_col_indices, key_col_indices+num_key_cols, key_cols_vect_out.begin(),
                   [&destination_table] (gdf_index_type const index) { return destination_table.get_column(index); });
   cudf::table key_col_sorted_table(key_cols_vect_out.data(), key_cols_vect_out.size());
+
+  // Copy the indices from device vector to gdf_column
+  rmm::device_vector<gdf_index_type> tmp = gdf_unique_indices(key_col_sorted_table, *context);
+  RMM_TRY(RMM_ALLOC(&unique_indices.data, sizeof(gdf_index_type)*tmp.size(), nullptr));
+  CUDA_TRY(cudaMemcpy(unique_indices.data, thrust::raw_pointer_cast(&tmp.front()), sizeof(gdf_index_type)*tmp.size(), cudaMemcpyDeviceToDevice));
```
Comment (jrhemstad) [inline cpp/src/groupby/new_groupby.cu]: This is inefficient. `gdf_unique_indices` should be updated to also return a `gdf_column`.

File: cpp/src/groupby/new_groupby.cu [hunk 4]
```diff
@@ -244,32 +246,37 @@ gdf_unique_indices(cudf::table const& input_table, gdf_context const& context)
...
     auto comp = row_equality_comparator<false>(*device_input_table, true);
     result_end = thrust::unique_copy(exec, counting_iter, counting_iter+nrows,
-                              unique_indices.data().get(),
+                              (gdf_index_type *)unique_indices.data,
                               comp);
   }
 
-  unique_indices.resize(thrust::distance(unique_indices.data().get(), result_end));
+  // size of the GDF column is being resized
+  RMM_TRY(RMM_REALLOC(&unique_indices.data, thrust::distance((gdf_index_type *)unique_indices.data, result_end)*sizeof(gdf_index_type), nullptr));
```
Comment (jrhemstad) [inline cpp/src/groupby/new_groupby.cu]: I believe that `RMM_REALLOC` is broken and does not preserve the original data.

File: cpp/src/groupby/new_groupby.cu [hunk 5]
```diff
@@ -244,32 +246,37 @@ gdf_unique_indices(cudf::table const& input_table, gdf_context const& context)
...
+  unique_indices.size = thrust::distance((gdf_index_type*)unique_indices.data, result_end);
```
Comment (jrhemstad) [inline cpp/src/groupby/new_groupby.cu]: Don't use C-style casts.

File: cpp/src/groupby/new_groupby.cu [hunk 6]
```diff
@@ -244,32 +246,37 @@ gdf_unique_indices(cudf::table const& input_table, gdf_context const& context)
...
+  unique_indices.dtype = cudf::gdf_dtype_of<gdf_index_type>();
 
   auto counting_iter = thrust::make_counting_iterator<gdf_size_type>(0);
   auto device_input_table = device_table::create(input_table);
   bool nullable = device_input_table.get()->has_nulls();
   if (nullable){
     auto comp = row_equality_comparator<true>(*device_input_table, true);
     result_end = thrust::unique_copy(exec, counting_iter, counting_iter+nrows,
-                              unique_indices.data().get(),
+                              (gdf_index_type *)unique_indices.data,
```
Comment (jrhemstad) [inline cpp/src/groupby/new_groupby.cu]: Don't use C-style casts 
```suggestion
                             static_cast<gdf_index_type *>(unique_indices.data),
```

File: cpp/src/groupby/new_groupby.cu [hunk 7]
```diff
@@ -244,32 +246,37 @@ gdf_unique_indices(cudf::table const& input_table, gdf_context const& context)
...
                               comp);
   } else {
     auto comp = row_equality_comparator<false>(*device_input_table, true);
     result_end = thrust::unique_copy(exec, counting_iter, counting_iter+nrows,
-                              unique_indices.data().get(),
+                              (gdf_index_type *)unique_indices.data,
```
Comment (jrhemstad) [inline cpp/src/groupby/new_groupby.cu]: C-style cast.

File: cpp/src/groupby/new_groupby.cu [hunk 8]
```diff
@@ -279,7 +286,7 @@ gdf_group_by_without_aggregations(cudf::table const& input_table,
   CUDF_EXPECTS(0 < num_key_cols, "number of key colums should be greater than zero");
 
   if (0 == input_table.num_rows()) {
-    return std::make_pair(cudf::table(), rmm::device_vector<gdf_index_type>());
+    return std::make_pair(cudf::table(), gdf_column {});
```
Comment (jrhemstad) [inline cpp/src/groupby/new_groupby.cu]: The returned column will not have the correct type here.

File: cpp/tests/groupby/without_agg/groupby_wo_agg_test.cu [hunk 9]
```diff
@@ -184,24 +184,28 @@ struct GroupByWoAggTest : public GdfTest {
...
+    gdf_column indices_col;
     EXPECT_NO_THROW(std::tie(output_table,
-                            indices_arr) = gdf_group_by_without_aggregations(input_table,
+                            indices_col) = gdf_group_by_without_aggregations(input_table,
                                                                             num_columns,
                                                                             groupby_col_indices.data(),
                                                                             &context));
 
-    copy_output_with_array(output_table.begin(), cpu_data_cols_out, indices_arr.data().get(), indices_arr.size(), this->cpu_out_indices);
+    copy_output_with_array(output_table.begin(), cpu_data_cols_out, (gdf_size_type *)indices_col.data, indices_col.size, this->cpu_out_indices);
```
Comment (jrhemstad) [inline cpp/tests/groupby/without_agg/groupby_wo_agg_test.cu]: C-style cast.

File: cpp/tests/groupby/without_agg/groupby_wo_agg_test.cu [hunk 10]
```diff
@@ -184,24 +184,28 @@ struct GroupByWoAggTest : public GdfTest {
...
 
     // Free results
+    RMM_TRY(RMM_FREE(indices_col.data, 0));
```
Comment (jrhemstad) [inline cpp/tests/groupby/without_agg/groupby_wo_agg_test.cu]: Use `gdf_column_free`

File: cpp/tests/groupby/without_agg/groupby_wo_agg_test.cu [hunk 11]
```diff
@@ -398,16 +402,19 @@ struct GroupValidTest : public GroupByWoAggTest<test_parameters>
...
+    gdf_column indices_col;
     EXPECT_NO_THROW(std::tie(output_table,
-                            indices_arr) = gdf_group_by_without_aggregations(input_table,
+                            indices_col) = gdf_group_by_without_aggregations(input_table,
                                                                             num_columns,
                                                                             groupby_col_indices.data(),
                                                                             &this->context));
     copy_output_with_array_with_nulls(
-            output_table.begin(), this->cpu_data_cols_out, this->cpu_data_cols_out_valid, indices_arr.data().get(), indices_arr.size(), this->cpu_out_indices);
+            output_table.begin(), this->cpu_data_cols_out, this->cpu_data_cols_out_valid, (gdf_size_type *)indices_col.data, indices_col.size, this->cpu_out_indices);
```
Comment (jrhemstad) [inline cpp/tests/groupby/without_agg/groupby_wo_agg_test.cu]: C-style cast.

File: cpp/tests/groupby/without_agg/groupby_wo_agg_test.cu [hunk 12]
```diff
@@ -398,16 +402,19 @@ struct GroupValidTest : public GroupByWoAggTest<test_parameters>
...
 
     // Free results
+    RMM_TRY(RMM_FREE(indices_col.data, 0));
```
Comment (jrhemstad) [inline cpp/tests/groupby/without_agg/groupby_wo_agg_test.cu]: `gdf_column_free`

File: cpp/src/groupby/new_groupby.cu (hunk 4 above)
Comment (rgsl888prabhu) [reply cpp/src/groupby/new_groupby.cu]: I did not see any issues in the test cases

Comment (rgsl888prabhu) [reply cpp/src/groupby/new_groupby.cu]: I can leave resizing if its actually broken but limit the address access with the size.

Comment (jrhemstad) [reply cpp/src/groupby/new_groupby.cu]: See https://github.com/rapidsai/rmm/issues/5

**Comment by rgsl888prabhu** [comment]
@jrhemstad

File: cpp/include/cudf/groupby.hpp [hunk 13]
```diff
@@ -191,10 +192,11 @@ gdf_unique_indices(cudf::table const& input_table, gdf_context const& context);
...
```
Comment (jrhemstad) [inline cpp/include/cudf/groupby.hpp]:
```suggestion
 *          - A non-nullable column of `GDF_INT32` elements containing the indices of the first occurrences of each unique row.
```

File: cpp/include/cudf/groupby.hpp [hunk 14]
```diff
@@ -191,10 +192,11 @@ gdf_unique_indices(cudf::table const& input_table, gdf_context const& context);
...
+ *            It can return only GDF_INT32 type and cannot contain null values.
```
Comment (jrhemstad) [inline cpp/include/cudf/groupby.hpp]:
```suggestion
```

**Comment by jrhemstad** [comment]
@rgsl888prabhu can you also fix https://github.com/rapidsai/cudf/issues/2335 in this PR?

**Comment by rgsl888prabhu** [comment]
> @rgsl888prabhu can you also fix #2335 in this PR? See the fix here: #2336

Fixed

**Pull request merged**